diff --git a/os_dep/osdep_service.c b/os_dep/osdep_service.c
index fd36e3b..07b3408 100644
--- a/os_dep/osdep_service.c
+++ b/os_dep/osdep_service.c
@@ -56,7 +56,7 @@ static ssize_t new_sync_read(struct file *filp, void __user *buf, __kernel_size_
         kiocb.ki_pos = *ppos;
         iov_iter_init(&iter, READ, &iov, 1, len);
 
-        ret = call_read_iter(filp, &kiocb, &iter);
+        ret = filp->f_op->read_iter(&kiocb, &iter);
         BUG_ON(ret == -EIOCBQUEUED);
         *ppos = kiocb.ki_pos;
         return ret;
