diff --git a/drivers/net/wireless/realtek/rtl8xxxu/core.c b/drivers/net/wireless/realtek/rtl8xxxu/core.c
index ee278f0548e4..a05d8f8249bc 100644
--- a/drivers/net/wireless/realtek/rtl8xxxu/core.c
+++ b/drivers/net/wireless/realtek/rtl8xxxu/core.c
@@ -5512,10 +5512,6 @@ static void rtl8xxxu_tx(struct ieee80211_hw *hw,
 		default:
 			break;
 		}
-		if (bmc && rtlvif && rtlvif->hw_key_idx != 0xff) {
-			tx_desc->txdw1 |= cpu_to_le32(TXDESC_EN_DESC_ID);
-			macid = rtlvif->hw_key_idx;
-		}
 	}
 
 	/* (tx_info->flags & IEEE80211_TX_CTL_AMPDU) && */